File: TwoPointer/3Sum/3Sum.kt
class Solution {
    fun threeSum(nums: IntArray): List<List<Int>> {
        nums.sort()                                    // in-place, stays IntArray
        val result = mutableListOf<List<Int>>()
        for(i in 0..<nums.size) {
            if (i > 0 && nums[i] == nums[i - 1]) continue
            twoSum(i, nums, result)
        }
        return result
    }

    fun twoSum(i: Int, input: IntArray, result: MutableList<List<Int>>): List<List<Int>> {
        var l = i+1
        var r = input.size-1 
        while(l<r) {
           val sum = input[i] + input[l] + input[r]
           if (sum < 0) {
            l++
           } else if (sum > 0) {
            r--
           } else {
            result.add(listOf(input[i], input[l], input[r]))
            l++
            r--
            while (l < r && input[l] == input[l - 1]) l++
           }
        }
        return result
    }
}

File: TwoPointer/3Sum/3Sum.swift
class Solution {
    func threeSum(_ nums: [Int]) -> [[Int]] {
        var result: [[Int]] = []
        let nums = nums.sorted()
        for i in 0..<nums.count {
            if i > 0 && nums[i] == nums[i-1] { continue }
            twoSum(i: i, nums: nums, result: &result)
        }
        return result
    }

    func twoSum(i: Int, nums: [Int], result: inout [[Int]]) {
        var l = i + 1
        var r = nums.count-1 
        while(l < r) {
            let sum = nums[i] + nums[l] + nums[r]
            if (sum < 0) {
                l+=1
            } else if (sum > 0) {
                r-=1
            } else {
                result.append([nums[i], nums[l], nums[r]])
                l+=1
                r-=1
                while(l < r && nums[l] == nums[l-1]) { l+=1 }
            }
        }
    }
}
